ykrt/src/compile/jitc_yk/codegen/x86_64/lsregalloc.rs

```diff
@@ -0,0 +1,1099 @@
+    /// Record the [InstIdx] of the last instruction that the value produced by an instruction is
+    /// used. By definition this must either be unused (if an instruction does not produce a value)
+    /// or `>=` the offset in this vector.
+    inst_vals_alive_until: Vec<InstIdx>,
+    /// Where on the stack is an instruction's value spilled? Set to `usize::MAX` if that offset is
+    /// currently unknown.
+    spilled_insts: Vec<usize>,
+    /// The abstract stack: shared between general purpose and floating point registers.
+    stack: AbstractStack,
+}
+
+impl<'a> LSRegAlloc<'a> {
+    pub(crate) fn new(m: &'a Module, inst_vals_alive_until: Vec<InstIdx>) -> Self {
+        #[cfg(debug_assertions)]
+        {
+            // We rely on the registers in GP_REGS being numbered 0..15 (inc.) for correctness.
+            for (i, reg) in GP_REGS.iter().enumerate() {
+                assert_eq!(reg.code(), u8::try_from(i).unwrap());
+            }
+
+            // We rely on the registers in FP_REGS being numbered 0..15 (inc.) for correctness.
+            for (i, reg) in FP_REGS.iter().enumerate() {
+                assert_eq!(reg.code(), u8::try_from(i).unwrap());
+            }
+        }
+
+        let mut gp_reg_states = [RegState::Empty; GP_REGS_LEN];
+        for reg in RESERVED_GP_REGS {
+            gp_reg_states[usize::from(reg.code())] = RegState::Reserved;
+        }
+
+        let mut fp_reg_states = [RegState::Empty; FP_REGS_LEN];
+        for reg in RESERVED_FP_REGS {
+            fp_reg_states[usize::from(reg.code())] = RegState::Reserved;
+        }
+
+        LSRegAlloc {
+            m,
+            gp_regset: RegSet::with_gp_reserved(),
+            gp_reg_states,
+            fp_regset: RegSet::with_fp_reserved(),
+            fp_reg_states,
+            inst_vals_alive_until,
+            spilled_insts: vec![usize::MAX; m.insts_len()],
+            stack: Default::default(),
+        }
+    }
+
+    /// Before generating code for the instruction at `iidx`, see which registers are no longer
+    /// needed and mark them as [RegState::Empty]. Calling this allows the register allocator to
+    /// use the set of available registers more efficiently.
+    pub(crate) fn expire_regs(&mut self, iidx: InstIdx) {
+        for reg in GP_REGS {
+            match self.gp_reg_states[usize::from(reg.code())] {
+                RegState::Reserved => {
+                    debug_assert!(self.gp_regset.is_set(reg));
+                }
+                RegState::Empty => {
+                    debug_assert!(!self.gp_regset.is_set(reg));
+                }
+                RegState::FromConst(_) => {
+                    debug_assert!(self.gp_regset.is_set(reg));
+                    // FIXME: Don't immediately expire constants
+                    self.gp_regset.unset(reg);
+                    self.gp_reg_states[usize::from(reg.code())] = RegState::Empty;
+                }
+                RegState::FromInst(reg_iidx) => {
```

Comment:
Is this regstate and `is_inst_var_still_used_at()` stuff needed because you do allocation at the same time as code-gen?

I'd imagined that an initial backwards pass would allocate all the registers before any code generation happens, then you don't need to keep asking "is the local still alive?" and "what's this register used for?".

Maybe I have an over-simplified view of this though.

Reply:
>Is this regstate and is_inst_var_still_used_at() stuff needed because you do allocation at the same time as code-gen?

No, both would be identical in that scenario.

What *would* be meaningfully different is that when doing everything in reverse you have the concept of "what's about to be used next". Right now, the register allocator knows the past but all it knows about the future is "when will this value last be used?". Different linear scan register allocators have various heuristics to decide which registers to reuse: we don't do anything clever yet though.